README.md
# Lambda Local Proxy

Lambda Local Proxy is a HTTP proxy that receives HTTP requests and invokes an AWS Lambda function. By combining with [lambci/docker-lambda](https://github.com/lambci/docker-lambda), you can test Lambda handler code locally or on CI without deploying.

This proxy can emulate following event formats at this moment:

* Application Load Balancer (ALB)
* Application Load Balancer (ALB) with multi-value headers enabled


## Usage

```
Usage of lambda-local-proxy:
  -e string
        Lambda API endpoint
  -f string
        Lambda function name (default "myfunction")
  -l string
        HTTP listen address (default any)
  -m    Enable multi-value headers. Effective only with -t alb
  -p int
        HTTP listen port (default 8080)
  -t string
        HTTP gateway type ("alb" for ALB) (default "alb")

  Environment variables:
    AWS_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_SESSION_TOKEN
```

Once it starts, access `http://localhost:8080` to invoke a lambda function.

Examples:

```bash
# Invoking local code through lambci/docker-lambda
$ docker run --rm \
  -e DOCKER_LAMBDA_WATCH=1 -e DOCKER_LAMBDA_STAY_OPEN=1 -p 9001:9001 \
  -v "$PWD":/var/task:ro,delegated \
  lambci/lambda:python3.8 lambda_function.lambda_handler
$ lambda-local-proxy -e http://localhost:9001 -t alb
```

```bash
# Invoking a deployed Lambda function
$ export AWS_REGION=...
$ export AWS_ACCESS_KEY_ID=...
$ export AWS_SECRET_ACCESS_KEY=...
$ lambda-local-proxy -t alb
```

## Development

### Build

```
$ go build
```

### Release

See [GoReleaser](https://goreleaser.com/) documents.

Cheat sheet:

```
# Build packages in ./dist
goreleaser --snapshot --skip-publish --rm-dist

# Release
git push         # make sure you pushed
git push --tags  # code and tags.
env GITHUB_TOKEN=... goreleaser
```

## License

Apache License, Version 2.0.
Copyright (C) 2020 Treasure Data Inc.



ChangeLog
2022-01-14 version 0.0.5:

* Bumped Golang, aws-sdk-go and aws-lambda-go
* Added ARM CPU build (`Darwin_arm64` and `Linux_arm64`), mainly to support Mac with Apple silicon

2021-05-07 version 0.0.4:

* Added `-l ADDRESS`(default: any) and `-p PORT` (default: `8080`) option

2021-05-05 version 0.0.3:

* Added RequestContext field to events with a dummy TargetGroupArn

2020-12-04 version 0.0.2:

* Added X-Forwarded-For, X-Forwarded-Proto, and X-Forwarded-Port headers

2020-11-17 version 0.0.1:

* The first release



alb.go
package main

import (
    "bytes"
    "encoding/base64"
    "encoding/json"
    "io"
    "mime"
    "net/http"
    "net/url"
    "strings"
    "github.com/aws/aws-lambda-go/events"
)

func NewALBPayloadBuilder(enableMultiValue bool) PayloadBuilder {
    return ALBPayloadBuilder{
        EnableMultiValue: enableMultiValue,
    }
}

type ALBPayloadBuilder struct {
    EnableMultiValue bool
}

func (pb ALBPayloadBuilder) BuildRequest(r *http.Request) ([]byte, error) {
    queryParams, err := url.ParseQuery(r.URL.RawQuery)
    if err != nil {
        return nil, err
    }

    body, isBase64Encoded, err := ReadBodyAsString(r)
    if err != nil {
        return nil, err
    }

    event := events.ALBTargetGroupRequest{
        HTTPMethod: r.Method,
        Path: r.URL.Path,
        Body: body,
        IsBase64Encoded: isBase64Encoded,
        RequestContext: events.ALBTargetGroupRequestContext{
            ELB: events.ELBContext{
                TargetGroupArn: "arn:aws:elasticloadbalancing:us-east-2:123456789012:targetgroup/lambda-local-proxy-dummy/1234567890123456",
            },
        },
    }

    if pb.EnableMultiValue {
        event.MultiValueQueryStringParameters = queryParams
        event.MultiValueHeaders = r.Header
    } else {
        event.QueryStringParameters = ArrayMapToFirstElementMap(queryParams)
        event.Headers = ArrayMapToFirstElementMap(r.Header)
    }

    return json.Marshal(event)
}

func (pb ALBPayloadBuilder) BuildResponse(blob []byte) (int, []byte, map[string][]string, error) {
    resp := events.ALBTargetGroupResponse{}

    err := json.Unmarshal(blob, &resp)

    var body []byte
    if err != nil {
        return BuildErrorResponse(err, "Invalid JSON response")
    } else if resp.IsBase64Encoded {
        binary, err := base64.StdEncoding.DecodeString(resp.Body)
        if err != nil {
            return BuildErrorResponse(err, "Invalid body in JSON response")
        }
        body = binary
    } else {
        body = []byte(resp.Body)
    }

    if pb.EnableMultiValue {
        return resp.StatusCode, body, resp.MultiValueHeaders, nil
    } else {
        return resp.StatusCode, body, MapToArrayMap(resp.Headers), nil
    }
}

func ReadBodyAsString(r *http.Request) (string, bool, error) {
    if IsTextContentType(r) {
        body, err := ReadFullyAsString(r.Body)
        return body, false, err
    } else {
        binary, err := ReadFullyAsBinary(r.Body)
        if err != nil {
            return "", false, err
        }
        body := base64.StdEncoding.EncodeToString(binary)
        return body, true, nil
    }
}

func ArrayMapToFirstElementMap(arrayMap map[string][]string) map[string]string {
    m := make(map[string]string)
    for key, values := range arrayMap {
        m[key] = values[0]
    }
    return m
}

func MapToArrayMap(m map[string]string) map[string][]string {
    arrayMap := make(map[string][]string)
    for key, value := range m {
        arrayMap[key] = []string { value }
    }
    return arrayMap
}

func ReadFullyAsBinary(reader io.Reader) ([]byte, error) {
    buf := new(bytes.Buffer)
    _, err := buf.ReadFrom(reader)
    return buf.Bytes(), err
}

func ReadFullyAsString(reader io.Reader) (string, error) {
    buf := new(bytes.Buffer)
    _, err := buf.ReadFrom(reader)
    return buf.String(), err
}

func IsTextContentType(r *http.Request) bool {
    contentType := r.Header.Get("Content-type")

    t, _, err := mime.ParseMediaType(contentType)
    if err != nil {
        return false
    }

    return strings.HasPrefix(t, "text/") || t == "application/json" || t == "application/javascript" || t == "application/xml"
}

func BuildErrorResponse(err error, message string) (int, []byte, map[string][]string, error) {
    body := "502 Bad Gateway\n" + message + "\n" + err.Error()
    return 502, []byte(body), make(map[string][]string), err
}



main.go
package main

import (
    "flag"
    "fmt"
    "log"
    "net/http"
    "os"
    "strings"
    "github.com/aws/aws-sdk-go/aws"
    "github.com/aws/aws-sdk-go/aws/session"
    "github.com/aws/aws-sdk-go/service/lambda"
)

type PayloadBuilder interface {
    BuildRequest(*http.Request) ([]byte, error)

    BuildResponse([]byte) (int, []byte, map[string][]string, error)
}

func main() {
    functionName := flag.String("f", "myfunction", "Lambda function name")
    bind := flag.String("l", "", "HTTP listen address (default any)")
    port := flag.Int("p", 8080, "HTTP listen port")
    endpoint := flag.String("e", "", "Lambda API endpoint")
    apiType := flag.String("t", "alb", "HTTP gateway type (\"alb\" for ALB)")
    albMultiValue := flag.Bool("m", false, "Enable multi-value headers. Effective only with -t alb")

    flag.Usage = func() {
        fmt.Println("Usage of lambda-local-proxy:")
        flag.PrintDefaults()
        fmt.Println("")
        fmt.Println("  Environment variables:")
        fmt.Println("    AWS_REGION, AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, AWS_SESSION_TOKEN")
    }
    flag.Parse()

    if *apiType != "alb" {
        fmt.Println("Unknown gateway type: " + *apiType)
        os.Exit(1)
    }

    pb := NewALBPayloadBuilder(*albMultiValue)
    client := MakeLambdaClient(*endpoint)
    handler := MakeInvokeLambdaHandler(client, *functionName, pb)

    http.HandleFunc("/", handler)

    listenAddress := fmt.Sprintf("%s:%d", *bind, *port)
    log.Fatal(http.ListenAndServe(listenAddress, nil))
}

func MakeLambdaClient(endpoint string) *lambda.Lambda {
    sess := session.Must(session.NewSessionWithOptions(session.Options{
        SharedConfigState: session.SharedConfigEnable,
    }))

    config := aws.Config{}
    if endpoint != "" {
        config.Endpoint = &endpoint
    }

    return lambda.New(sess, &config)
}

func MakeInvokeLambdaHandler(client *lambda.Lambda, functionName string, pb PayloadBuilder) func(http.ResponseWriter, *http.Request) {
    return func(w http.ResponseWriter, r *http.Request) {
        // Add proxy headers
        r.Header.Add("X-Forwarded-For", r.RemoteAddr[0:strings.LastIndex(r.RemoteAddr, ":")])
        r.Header.Add("X-Forwarded-Proto", "http")
        r.Header.Add("X-Forwarded-Port", "8080")

        // Parse HTTP response and create an event
        payload, err := pb.BuildRequest(r)
        if err != nil {
            WriteErrorResponse(w, "Invalid request", err)
            return
        }

        // Invoke Lambda with the event
        output, err := client.Invoke(&lambda.InvokeInput{
            FunctionName: aws.String(functionName),
            Payload: payload,
        })
        if err != nil {
            WriteErrorResponse(w, "Failed to invoke Lambda", err)
            return
        }
        if output.FunctionError != nil {
            WriteErrorResponse(w, "Lambda function error: " + *output.FunctionError, nil)
            return
        }

        // Build a response
        status, body, headers, err := pb.BuildResponse(output.Payload)
        if err != nil {
            WriteErrorResponse(w, "Invalid JSON response", err)
            return
        }

        // Write the response - headers, status code, and body
        for key, values := range headers {
            for _, value := range values {
                w.Header().Add(key, value)
            }
        }
        w.WriteHeader(status)
        w.Write(body)
        return
    }
}

func WriteErrorResponse(w http.ResponseWriter, message string, err error) {
    body := "502 Bad Gateway\n" + message
    if err != nil {
        body += "\n" + err.Error()
    }
    w.WriteHeader(502)  // Bad Gateway
    w.Write([]byte(body))
    return
}
